ocular/src/keyring.rs

```diff
@@ -0,0 +1,752 @@
+        }
+
+        // Check if new key exists
+        if self.key_exists(new_name)? && !override_if_exists {
+            eprintln!("New key name '{}', already exists.", new_name);
+            return Err(KeyStoreError::Exists(new_name.to_string()));
+        }
+
+        // Proceed with rename
+        let current_name = &KeyName::new(current_name)
+            .unwrap_or_else(|_| panic!("Could not create KeyName for '{}'.", current_name));
+        let new_name = &KeyName::new(new_name)
+            .unwrap_or_else(|_| panic!("Could not create KeyName for '{}'.", new_name));
+
+        let key = self
+            .key_store
+            .as_ref()
+            .expect("Error accessing key store.")
+            .load(current_name)
+            .expect("Could not load key.");
+
+        // Create new key.
+        let _result = self
+            .key_store
+            .as_ref()
+            .expect("Error accessing key store.")
+            .store(new_name, &key);
+
+        // Delete old key.
+        let _result = self.delete_key(current_name);
+
+        Ok(())
+    }
+
+    fn get_key_address(&self, name: &str) -> Result<PublicKeyOutput, KeyStoreError> {
+        // Check if key exists
+        if !self.key_exists(name)? {
+            eprintln!("Key '{}', does not exist.", name);
+            return Err(KeyStoreError::DoesNotExist(name.to_string()));
+        }
+
+        // Prepare key and parse into pem
+        let key_name = &KeyName::new(name)
+            .unwrap_or_else(|_| panic!("Could not create KeyName for '{}'.", name));
+        let key = self
+            .key_store
+            .as_ref()
+            .expect("Error accessing key store.")
+            .load(key_name)
+            .expect("Could not load key");
+        let key: k256::elliptic_curve::SecretKey<k256::Secp256k1> = key
+            .to_pem(LineEnding::default())
+            .expect("Could not convert to pem.")
+            .parse()
+            .expect("Could not parse pem");
+
+        // Convert to deepspace private key for address conversion
+        let key_bytes = key.to_bytes();
+
+        // Obtain public key bytes
+        let secp256k1_engine = secp256k1::Secp256k1::new();
+        let secp256k1_secret_key =
+            secp256k1::SecretKey::from_slice(&key_bytes).expect("Could not create secret key.");
+        let secp256k1_public_key =
+            secp256k1::PublicKey::from_secret_key(&secp256k1_engine, &secp256k1_secret_key);
+        let secp256k1_public_key_bytes = secp256k1_public_key.serialize();
+
+        // Obtain address hash
+        let sha256 = Sha256::digest(secp256k1_public_key_bytes);
+        let ripemd160 = Ripemd::digest(&sha256);
+        let mut bytes: [u8; 20] = Default::default();
+        bytes.copy_from_slice(&ripemd160[..]);
+
+        // Finally encode address to bech32
+        // TODO: Support other prefixes
+        let bech32_encoded_address =
```

Comment:
It doesn't look like you have `cosmrs` as a dependency yet, but it can compute all of this for you. If you compute the `verifying_key` above, then you can do:

```suggestion
        let account_id = cosmrs::PublicKey::from(verifying_key).account_id(COSMOS_ADDRESS_PREFIX)?;
```

This will return a [`cosmrs::AccountId`](https://docs.rs/cosmrs/latest/cosmrs/struct.AccountId.html), which is a newtype around a Bech32-encoded string. You can access the contents as a `&str` with `as_ref()`.

Reply:
Indeed, deep_space can also provide this functionality. However based off of a previous conversation with @cbrit, I believe we want this library to refrain from wrapping around other cosmos libraries (though I suppose the sdk might be an exception). 

There's certainly value in using what already works in other libs (especially the sdk), but I do also see the value of reimplementing things at a lower level to prevent this from becoming a wrapper around a wrapper.

Reply:
It sounds like you'll be duplicating a large amount of functionality that's already implemented in CosmRS. That seems pretty bad to me?

I thought the point of Ocular was to build something higher-level on top of CosmRS. The whole point of setting up https://github.com/cosmos/cosmos-rust was to have a single place where we could collaborate on common Cosmos Rust functionality instead of duplicating things over and over and over again, but it sounds like Ocular is doing just that?

Reply:
That's fair. I'm unfamiliar with the context around the specific desired distinction between the 2 repositories. We already have a handful of use cases we want ocular for. Would there be any benefit of bringing (full or partial) Ocular functionalities into cosmrs at that point since some of them might be relatively common (e.g. key store/ring management)?

Reply:
An integration between CosmRS and Signatory would probably be a good idea if we can get the other stakeholders to agree to it

Reply:
Thanks @tony-iqlusion. We definitely want cosmrs as as dependency.

Reply:
Updated bech32 encoding flow.